### PR TITLE
chore(sdk): remove python 3.6 dead code

### PR DESCRIPTION
**Pull request by connor-mccarthy**
This is a follow-up PR addressing a [comment](https://github.com/kubeflow/pipelines/pull/7428#discussion_r829542363) from @chensun on #7428.

**Description of changes**
- migrate modules to >= python 3.7
- update test images to greatest supported python version
- remove unused import

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/compiler/compiler_test.py [hunk 1]
```diff
@@ -553,7 +552,7 @@ def test_use_task_final_status_in_non_exit_op_yaml(self):
 - {name: message, type: PipelineTaskFinalStatus}
 implementation:
   container:
-    image: python:3.7
+    image: python:3.9
```
Comment (chensun) [inline sdk/python/kfp/compiler/compiler_test.py:555]: I don't think we need to change this. Our default image for lightweight component is still Python:3.7

File: sdk/python/kfp/components/structures.py [hunk 2]
```diff
@@ -305,12 +305,10 @@ class ComponentSpec(BaseModel):
         implementation: The implementation of the component. Either an executor
             (container, importer) or a DAG consists of other components.
     """
-    # TODO(ji-yaqi): Update to OrderedDict for inputs and outputs once we drop
-    # Python 3.6 support
```
Comment (chensun) [inline sdk/python/kfp/components/structures.py:309]: I forget the context on this TODO. Doesn't `Dict` guarantee order already?
Do check with @ji-yaqi though.

**Review by chensun — COMMENTED** [review]
Title is a bit misleading, as we dropped Python 3.6 support via https://github.com/kubeflow/pipelines/pull/7303.
This is more of some dead code cleanup, I think. Maybe reframe the title a bit?

File: sdk/python/kfp/components/structures.py (hunk 2 above)
Comment (connor-mccarthy) [reply sdk/python/kfp/components/structures.py:309]: Good catch, @chensun. I think you're right... explanation:

Dicts are officially ordered in >= Python 3.7. This is a nuanced change: Python `dict`s are ordered, but are not `OrderedDict`s. We accept `dict` (e.g., dictionaries without [the special `OrderedDict` methods](https://docs.python.org/3/library/collections.html#collections.OrderedDict)), so it seems like we should not needlessly narrow this type from the superclass `dict` to the subclass `typing.OrderedDict`.

In code:
```python
import typing
import collections

d = {"a": True, "b": False}
od = collections.OrderedDict({"a": True, "b": False})

assert isinstance(od, dict) # type-checking would pass with the looser type
assert not isinstance(d, typing.OrderedDict) # type-checking would fail with the narrower type
```

@ji-yaqi, do you agree? If so, I will revert.

Further reading: [Choosing Between OrderedDict and dict](https://realpython.com/python-ordereddict/#choosing-between-ordereddict-and-dict)

**Comment by connor-mccarthy** [comment]
> Title is a bit misleading, as we dropped Python 3.6 support via #7303. This is more of some dead code cleanup, I think. Maybe reframe the title a bit?

Thanks @chensun. Updated.

File: sdk/python/kfp/compiler/compiler_test.py (hunk 1 above)
Comment (connor-mccarthy) [reply sdk/python/kfp/compiler/compiler_test.py]: Good call. I misunderstood this template definition as having a functional effect for our tests.

File: sdk/python/kfp/compiler_cli_tests/test_data/pipeline_with_task_final_status_yaml.json [hunk 3]
```diff
@@ -66,7 +66,7 @@
             "pipeline status:",
             "{{$.inputs.parameters['status']}}"
           ],
-          "image": "python:3.7"
+          "image": "python:3.9"
```
Comment (chensun) [inline sdk/python/kfp/compiler_cli_tests/test_data/pipeline_with_task_final_status_yaml.json:69]: revert this change also? (including the .py and .json file)

Comment (connor-mccarthy) [reply sdk/python/kfp/compiler_cli_tests/test_data/pipeline_with_task_final_status_yaml.json]: Done

File: sdk/python/kfp/components/structures.py (hunk 2 above)
Comment (connor-mccarthy) [reply sdk/python/kfp/components/structures.py:309]: Updated: I reverted from `Dict` to `OrderedDict` in [b72d22f](https://github.com/kubeflow/pipelines/pull/7440/commits/b72d22f72d1c974ba564049d8065b77183910201).

Please let me know if this is not desired.

Comment (ji-yaqi) [reply sdk/python/kfp/components/structures.py:309]: OrderedDict SGTM, thank you!

**Review by ji-yaqi — COMMENTED** [review]
/lgtm
/approve

Thank you!
